File: core/Decoder.py
import cv2
import numpy as np

class Decoder:
    def __init__(self, receiver, img_tool, exporter, tracker):
        self._receiver = receiver
        self._img_tool = img_tool
        self._exporter = exporter
        self._tracker = tracker
        self._last_frame = None
        self._ref_frame = None
        self._ref_landmarks = None
        self._mask = None

    def start_reconstruction(self, output_name, fps, width, height):
        self._exporter.setup(output_name, fps, width, height)
        print(f"[Decoder] Starting hybrid reconstruction (Reference-Generated-Reference)...")
        
        count = 0
        try:
            while True:
                payload = self._receiver.receive_data()
                if payload is None: break
                
                count += 1
                if payload["type"] == "FULL":
                    frame = self._img_tool.from_base64(payload["data"])
                    mask_img = self._img_tool.from_base64(payload["fringe_mask"])
                    if frame is not None:
                        self._ref_frame = frame.copy()
                        self._ref_landmarks = self._tracker.deserialize_landmarks(payload["landmarks"])
                        self._mask = mask_img
                        # Write reference frame
                        self._exporter.write_frame(frame)
                        self._last_frame = frame
                
                elif payload["type"] == "MOTION":
                    if self._ref_frame is not None:
                        target_lm = self._tracker.deserialize_landmarks(payload["target_landmarks"])
                        # Generate frame using semantic motion
                        gen = self._img_tool.apply_hybrid_motion(
                            self._ref_frame, self._ref_landmarks, target_lm, self._mask, payload["boundary_vec"]
                        )
                        # Smoothly blend with previous frame to enhance visual consistency
                        if self._last_frame is not None:
                            final = cv2.addWeighted(gen, 0.7, self._last_frame, 0.3, 0)
                            self._exporter.write_frame(final)
                            self._last_frame = final
                        else:
                            self._exporter.write_frame(gen)
                            self._last_frame = gen

                if count % 100 == 0:
                    print(f"\rProcessing frame {count}...", end="", flush=True)
        finally:
            self._exporter.release()

File: semcodec/src/core/Decoder.ts
import type { SemCodecPayload, DecoderStats } from '../types/payloads';
import type { SemCodecConfig, ResolvedConfig } from '../types/config';
import { resolveConfig } from '../utils/config';
import { FaceTracker } from '../vision/FaceTracker';
import { ImageBlender } from '../vision/ImageBlender';
import type { NormalizedLandmark } from '@mediapipe/tasks-vision';

export type OnDecoderStatsCallback = (stats: DecoderStats) => void;

export class Decoder {
  private readonly config: ResolvedConfig;
  private tracker: FaceTracker;
  private blender: ImageBlender;
  
  private refFrame: ImageData | null = null;
  private refLandmarks: NormalizedLandmark[] | null = null;
  private fringeMask: ImageData | null = null;
  private lastFrame: ImageData | null = null;

  private initialized = false;
  private onStats?: OnDecoderStatsCallback;

  constructor(config?: SemCodecConfig) {
    this.config = resolveConfig(config);
    this.tracker = new FaceTracker();
    this.blender = new ImageBlender(1, 1, this.config);
  }

  /**
   * Initializes the decoder by loading MediaPipe WASM models.
   * Required for FaceTracker deserialization mapping if needed.
   */
  async initialize(): Promise<void> {
    await this.tracker.initialize(this.config);
    this.initialized = true;
  }

  /**
   * Decodes a received payload into a displayable frame.
   *
   * @param payload - A SemCodecPayload received from the network.
   * @returns Reconstructed ImageData ready to draw to a canvas.
   */
  async decode(payload: SemCodecPayload): Promise<ImageData | null> {
    if (!this.initialized) throw new Error('Decoder not initialized. Call initialize() first.');
    
    const t0 = performance.now();

    if (payload.type === 'FULL') {
      const frame = await this.blender.decodeFromBase64(payload.frame, 'image/jpeg');
      const mask = await this.blender.decodeFromBase64(payload.fringeMask, 'image/png');
      
      this.refFrame = frame;
      this.refLandmarks = this.tracker.deserializeLandmarks(payload.landmarks);
      this.fringeMask = mask;
      this.lastFrame = frame;
      
      this.emitStats(payload.frameIndex, payload.type, performance.now() - t0);
      return frame;
    }

    if (payload.type === 'MOTION') {
      if (!this.refFrame || !this.refLandmarks || !this.fringeMask) {
        // We received a motion payload but don't have a reference frame yet.
        // Wait for the next FULL keyframe.
        return null;
      }
      
      const targetLandmarks = this.tracker.deserializeLandmarks(payload.targetLandmarks);
      const generated = this.blender.applyHybridMotion(
        this.refFrame, 
        this.refLandmarks, 
        targetLandmarks,
        this.fringeMask, 
        payload.boundaryVector
      );

      // Temporal blending with previous frame for visual consistency
      const final = this.blendWithPrev(generated, this.lastFrame);
      this.lastFrame = final;
      
      this.emitStats(payload.frameIndex, payload.type, performance.now() - t0);
      return final;
    }

    return null;
  }

  private blendWithPrev(current: ImageData, prev: ImageData | null): ImageData {
    if (!prev) return current;
    
    const w = this.config.blendWeight;
    if (w >= 1.0) return current;
    if (w <= 0.0) return prev;
    
    const out = new ImageData(current.width, current.height);
    for (let i = 0; i < current.data.length; i++) {
      // Alpha channel remains 255
      if ((i + 1) % 4 === 0) {
          out.data[i] = 255;
          continue;
      }
      out.data[i] = (current.data[i]! * w + prev.data[i]! * (1 - w));
    }
    return out;
  }

  /**
   * Registers a callback to receive per-frame decoding statistics.
   */
  onFrameStats(callback: OnDecoderStatsCallback): this {
    this.onStats = callback;
    return this;
  }

  private emitStats(frameIndex: number, type: 'FULL'|'MOTION', time: number) {
      this.onStats?.({
          frameIndex,
          payloadType: type,
          processingTimeMs: time
      });
  }

  /**
   * Resets decoder state. Call when reconnecting or switching stream.
   */
  reset(): void {
    this.refFrame = null;
    this.refLandmarks = null;
    this.fringeMask = null;
    this.lastFrame = null;
  }
}
